shell/src/stats/memory.rs

```diff
@@ -0,0 +1,221 @@
+use std::fmt;
+use std::fs::File;
+use std::io::prelude::*;
+use std::path::Path;
+use std::process::Command;
+use core::str::{from_utf8};
+
+pub type MemoryStatsResult<T> = std::result::Result<T, MemoryStatsError>;
+
+#[derive(Serialize, PartialEq, Clone, Debug)]
+pub struct LinuxData {
+    page_size: usize,
+    size: String,
+    resident: String,
+    shared: String,
+    text: String,
+    lib: String,
+    data: String,
+    dt: String
+}
+
+#[derive(Serialize, PartialEq, Clone, Debug)]
+pub struct DarwinOsData {
+    page_size: usize,
+    mem: f64,
+    resident: String
+}
+
+#[derive(Serialize,PartialEq, Clone, Debug)]
+#[serde(untagged)]
+pub enum MemoryData {
+    Linux(LinuxData),
+    DarwinOs(DarwinOsData)
+}
+
+impl From<LinuxData> for MemoryData {
+    fn from(data: LinuxData) -> Self {
+        MemoryData::Linux(data)
+    }
+}
+
+impl From<DarwinOsData> for MemoryData {
+    fn from(data: DarwinOsData) -> Self {
+        MemoryData::DarwinOs(data)
+    }
+}
+
+#[derive(PartialEq, Debug)]
+pub enum MemoryStatsError {
+    InvalidPid,
+    FileRead,
+    ParsingData,
+    NotSupportedOs
+}
+
+impl fmt::Display for MemoryStatsError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str((self as &dyn error::Error).description())
+    }
+}
+
+impl error::Error for MemoryStatsError {
+    fn description(&self) -> &str {
+        match *self {
+            MemoryStatsError::InvalidPid => "MemoryStatsError: Invalid process pid",
+            MemoryStatsError::FileRead => "MemoryStatsError: Could not read stats file",
+            MemoryStatsError::ParsingData => "MemoryStatsError: Error parsing data",
+            MemoryStatsError::NotSupportedOs => "MemoryStatsError: Not supported OS",
+        }
+    }
+}
+
+lazy_static! {
+    static ref LINUX_DATA: Regex = Regex::new(r"^(?P<size>\d+)\s+(?P<resident>\d+)\s+(?P<shared>\d+)\s+(?P<text>\d+)\s+(?P<lib>\d+)\s+(?P<data>\d+)\s+(?P<dt>\d+).*").expect("Invalid regex");
+    static ref MAC_DATA: Regex = Regex::new(r"\d+\s+(?P<mem>[0-9.]+)\s+(?P<resident>\d+)").expect("Invalid regex");
+}
+
+pub struct Memory {
+    pub pid: i32,
+    pub page_size: usize
+}
+
+impl Memory {
+    pub fn new(opt_pid: Option<i32>) -> Self {
+        Memory {
+            pid: Memory::get_pid(opt_pid),
+            page_size: page_size::get()
+        }
+    }
+
+    pub fn get_memory_stats(&self) -> MemoryStatsResult<MemoryData> {
+        if cfg!(target_os = "linux") {
+            self.get_linux_data( format!("/proc/{}/statm", self.pid) )
+        } else if cfg!(target_os = "macos") || cfg!(target_os = "ios") {
+            self.get_mac_data()
+        } else {
+            Err(MemoryStatsError::NotSupportedOs)
+        }
+    }
+
+    /// private helpers
+
+    fn get_pid(opt_pid: Option<i32>) -> i32 {
+        opt_pid.unwrap_or(nix::unistd::getpid().as_raw())
+    }
+
+    fn get_mac_data(&self) -> MemoryStatsResult<MemoryData> {
+        let output = Command::new("sh")
```

Comment:
I believe it should be possible to execute `ps` without wrapping `sh -c` over it.

Reply:
Done. Lesson learned: do not use apostrophes or spaces in args, they will be escaped and command will not work